### PR TITLE
[BEAM-961] Add starting number to CountingInput

### PR DESCRIPTION
**Pull request by vladisav**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
R: @jbonofre
CC: @davorbonaci 
CC: @dhalperi

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java [hunk 1]
```diff
@@ -35,22 +35,32 @@
...
  * <pre>{@code
  * Pipeline p = ...
  * PTransform<PBegin, PCollection<Long>> producer = CountingInput.upTo(1000);
  * PCollection<Long> bounded = p.apply(producer);
  * }</pre>
  *
+ * <p>To produce a bounded {@code PCollection<Long>} starting from {@code startOffset},
+ * use {@link CountingInput#forSubrange(long, long)} :
+ *
+ * <pre>{@code
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java]: This code example is probably unnecessary - can just say at line 56, "use {@link ...} instead".

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java [hunk 2]
```diff
@@ -76,6 +86,16 @@ public static BoundedCountingInput upTo(long numElements) {
   }
 
   /**
+   * Creates a {@link BoundedCountingInput} that will produce the specified number of elements,
+   * starting from {@code startOffset} to (excluding) {@code endOffset}.
+   */
+  public static BoundedCountingInput forSubrange(long startOffset, long endOffset) {
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java]: I think "index" might be a better name than "offset" here - offsets are usually measured relative to something important (e.g. the beginning of a file).

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java [hunk 3]
```diff
@@ -76,6 +86,16 @@ public static BoundedCountingInput upTo(long numElements) {
   }
 
   /**
+   * Creates a {@link BoundedCountingInput} that will produce the specified number of elements,
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingInput.java]: This overload does not let you specify the number of elements.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java [hunk 4]
```diff
@@ -83,6 +83,20 @@
   }
 
   /**
+   * Creates a {@link BoundedSource} that will produce the specified number of elements,
+   * from {@code startOffset} to {@code numElements - 1}.
+   *
+   * @deprecated use {@link CountingInput#forSubrange(long, long)} instead
+   */
+  @Deprecated
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java]: Why add a method and immediately deprecate it? If you want users to not use it, package-local visibility is sufficient for that.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java [hunk 5]
```diff
@@ -83,6 +83,20 @@
   }
 
   /**
+   * Creates a {@link BoundedSource} that will produce the specified number of elements,
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java]: Same nitpicks about this comment.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java [hunk 6]
```diff
@@ -83,6 +83,20 @@
...
+   * from {@code startOffset} to {@code numElements - 1}.
+   *
+   * @deprecated use {@link CountingInput#forSubrange(long, long)} instead
+   */
+  @Deprecated
+  static BoundedSource<Long> createSourceForSubrange(long startOffset, long endOffset) {
+    checkArgument(endOffset > startOffset, "numElements (%s) must be greater than 0",
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CountingSource.java]: It's better to mention in an error message the things that the user directly specifies. Referring to numElements implies that the user specified a parameter called numElements, which is not the case. Rephrase as "endOffset must be greater than startOffset"?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/CountingSourceTest.java [hunk 7]
```diff
@@ -79,6 +79,26 @@ public static void addCountingAsserts(PCollection<Long> input, long numElements)
       .isEqualTo(numElements - 1);
   }
 
+  public static void addCountingAsserts(PCollection<Long> input, long start, long end) {
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CountingSourceTest.java]: Why this duplication of tests: what's the difference between CountingSourceTest and CountingInputTest?

Comment (vladisav) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CountingSourceTest.java]: You're right, it's the same. I also refactored&removed some duplicate code from CountingInputTest

**Comment by jkff** [comment]
Looks good as far as I'm concerned; I'll let JB do the rest of the review, thanks!

**Comment by dhalperi** [comment]
Minor drive-by comments:

* Why only add this support to the bounded source? Presumably the unbounded source would benefit from the same change. In general, we strive for unified APIs where we can.
* I probably would have added a test where start index (and/or end index) was negative, to sanity check those cases.

**Pull request closed without merging**